### PR TITLE
added alert if it's between total and fee

### LINKED ISSUE #467: User has balance in Wallet but receives message of insufficient funds for deposit
Similar to #466, when doing a deposit, there should be an estimation of the fee and where it is paid rather than just saying that there are no funds (the user is seeing that he has some funds)

WechatIMG351 - 小哥榛果

### PR DESCRIPTION
**Pull request by adrifdez**
Closes #467 

### What does this PR does?

Add an alert in build transaction step if the amount is between max amount and maxamountwithfee and show the fee in eth below the button
image

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Respect code style and lint
- [ ] Update documentation (if needed)

### CODE REVIEW

File: src/utils/accounts.ts
Comment (adrifdez) [reply src/utils/accounts.ts:55]: Yes better. We are still supporting prices from coordinator in case that price updater fails but in the future

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (adrifdez) [reply src/views/shared/amount-input/amount-input.view.jsx:25]: The one that you says is `isAmountMoreThanFunds` this one is more `isAmountLessThanFundsButMoreWithFees`, could we change to that?

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (adrifdez) [reply src/views/shared/amount-input/amount-input.view.jsx:99]: Yes I was keeping that part to not break other things but I think that we can refactor that without problems

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (adrifdez) [reply src/views/shared/amount-input/amount-input.view.jsx:101]: Yes I think that this check is old because is not posible to introduce negative values and it's better check newAmount. I kept this check because I don't know if it's posible to reach to that step with a negative value. Could do you think important refactor this part removing that logic or just change to newAmount?

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (adrifdez) [reply src/views/shared/amount-input/amount-input.view.jsx:105]: This show an alert on the parent component and send a signal that the amount is invalid and  `isAmountLessThanFundsButMoreWithFees` 
The behaivour of `newAmountWithFee` is valid for every TxType

File: src/views/transaction/components/fee/fee.view.jsx
Comment (adrifdez) [reply src/views/transaction/components/fee/fee.view.jsx:44]: I tried in the beginning but I found other alert showing the whole number including more decimals than 18, I don't know if we can round when are absolute values related to fees
@AlbertoElias could you shed us some light about this issue?

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (amonsosanz) [reply src/views/shared/amount-input/amount-input.view.jsx:25]: Right. Here we store if the TX cannot be done because **due to the fee** we are exceeding the funds. What about `areFundsExceededDueToFee`. I think this name clearly explains the point.

I would also consider renaming `isAmountMoreThanFunds` to `isAmountWithFeeMoreThanFunds`. Names are very important and this one is not very precise.

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (amonsosanz) [reply src/views/shared/amount-input/amount-input.view.jsx:99]: I think currently this code is a bit difficult to follow due to some not very well-chosen names. I'm of the opinion to take any opportunity to do this kind of refactors that help in making the code simple to follow.

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (amonsosanz) [reply src/views/shared/amount-input/amount-input.view.jsx:101]: No, I would not remove this logic. Here we still have to check for negative values from the user. The regex is an extra security measure on the input, but we should not rely only on input validation.

All we need to do IMO is check the right value here, which is what the user inputs (`newAmount`) and not what the user inputs plus the fee (`newAmountWithFee`).

Checking if `newAmountWithFee` is negative is a bug.

File: src/views/shared/amount-input/amount-input.view.jsx
Comment (amonsosanz) [reply src/views/shared/amount-input/amount-input.view.jsx:105]: I still don't get why this logic only applies to a `Deposit`. I mean, why we don't want to benefit from this feature when we are doing a withdraw or exit? Like here, where it "looks like" I do have enough funds, but still get an error because of the fee. Wouldn't it be nice to let the user know the fee is the reason as we do in the deposit?

image

Comment (adrifdez) [reply src/views/shared/amount-input/amount-input.view.jsx:105]: I think that this feature was requested just for deposit and `AmountInput` is  just taking in account the L2 fee in a exit Tx. For your approach we have something similar that is checking if you have enough eth amount in a exit
image

Comment (adrifdez) [reply src/views/shared/amount-input/amount-input.view.jsx:105]: But we can remove that conditional because that logic just affect if the parent is listening `areFundsExceededDueToFee` and maybe in the future some new feature consume it

Comment (amonsosanz) [reply src/views/shared/amount-input/amount-input.view.jsx]: mmm... interesting, then I don't get why I'm not seeing this msg in the capture above...

Ok, anyway, if this was planned only for Deposits I close this one. Thx!

File: src/views/transaction/components/fee/fee.view.jsx
Comment (amonsosanz) [reply src/views/transaction/components/fee/fee.view.jsx:44]: IMO this is excessive precision, especially for a couple of estimations:

image

**Review by amonsosanz — APPROVED** [review]

File: src/views/transaction/components/fee/fee.view.jsx
Comment (adrifdez) [reply src/views/transaction/components/fee/fee.view.jsx:44]: Yes it's showing more decimals that token allow and an estimation of eth fees that not should be below 1 cent in a long time

Comment (AlbertoElias) [reply src/views/transaction/components/fee/fee.view.jsx:44]: In this function specifically, we are talking about the fees in fiat. Which means it should be fixed to 2 decimals. Regarding token values, they should be fixed to 6 decimals. Please check views like Tx Details to see how we are handling it elsewhere

File: src/views/transaction/components/transaction-form/transaction-form.view.jsx
Comment (AlbertoElias) [inline src/views/transaction/components/transaction-form/transaction-form.view.jsx:298]: The fees for Deposits are gas fees, not L2 fees. feesTask data is for L2 fees

File: src/views/transaction/components/fee/fee.view.jsx
Comment (adrifdez) [reply src/views/transaction/components/fee/fee.view.jsx:44]: Should we fix to 2 decimals `getAmountInPreferredCurrency`?

File: src/views/transaction/components/transaction-form/transaction-form.view.jsx
Comment (AlbertoElias) [inline src/views/transaction/components/transaction-form/transaction-form.view.jsx:70]: Why do you need to set it instead of just using `estimatedDepositFeeTask` directly?

File: src/views/transaction/components/fee/fee.view.jsx
Comment (AlbertoElias) [reply src/views/transaction/components/fee/fee.view.jsx]: Might be worth creating an issue in Zenhub to refactor that. But I wouldn't do it in this PR. Can you take care of creating the issue and patching it manually in this pr?

File: src/views/transaction/components/transaction-form/transaction-form.view.jsx
Comment (adrifdez) [reply src/views/transaction/components/transaction-form/transaction-form.view.jsx:70]: To check status just one time and not introduce logic inside `AmountInput`

File: src/views/transaction/components/fee/fee.view.jsx
Comment (AlbertoElias) [reply src/views/transaction/components/fee/fee.view.jsx]: Waiting for the commit that fixes to 2 decimals in this pr. We can't ship it if it shows too many decimals

Comment (adrifdez) [reply src/views/transaction/components/fee/fee.view.jsx]: Really weird. I changed that in this commit https://github.com/hermeznetwork/wallet-ui/pull/665/commits/b27c12e938c0445441aed231f42b88685e8aafd8

Comment (adrifdez) [reply src/views/transaction/components/fee/fee.view.jsx]: ahh you mean fixed here

Comment (adrifdez) [reply src/views/transaction/components/fee/fee.view.jsx]: updated

**Review by AlbertoElias — APPROVED** [review]

**Pull request merged**
